### ISSUE TITLE
Holodeck vanishes without an error message

### ISSUE THREAD

**Issue by NancyFulda** [BYU-PCCL/holodeck-engine#30]
I've appended the code I'm executing below. Holodeck dies mysteriously after learning iteration 108.

(The reinforce package is from the pytorch examples Github repo with two lines changed:
        self.affine1 = nn.Linear(4, 128) ==> self.affine1 = nn.Linear(3, 128)
         self.affine2 = nn.Linear(128, 2) ==> self.affine2 = nn.Linear(128, 81)
)

--Nancy

import numpy as np
import pickle

from Holodeck import Holodeck, Agents
from Holodeck.Environments import HolodeckEnvironment
from Holodeck.Sensors import Sensors

import pygame

import sys
print(sys.version)
assert sys.version_info >= (3,5) #This script only works with python3

HEADLESS = True

import reinforce 

commands = []
for a in range(3):
    for b in range(3):
        for c in range(3):
            for d in range(3):
                commands.append(np.array([0.5*a-0.5, 0.5*b-0.5, 0.5*c-0.5, 5*d]))   

def get_vector(raw_command):
    return commands[raw_command]

def learn_to_hold_position(screen):
    #This function trains the drone to recover from random
    #impulses by learning to minimize the image deltas from
    #its camera

    env = Holodeck.make("UrbanCity")

    NUM_TRIALS = 10000
    for e in range(NUM_TRIALS): #number of trials
        avg_velocity_cost = 0
        env.reset()
    
        black = 0, 0, 0
        if HEADLESS != True:
            screen.fill(black)
            pygame.display.flip()

        # rise to elevation
        command = np.array([0, 0, 0, 10])
        for _ in range(100):
            state, reward, terminal, _ = env.step(command)

        DIFFICULTY = 3
        command = (np.random.random(4) - 0.5) * DIFFICULTY
        command[3] = 10
        #print(command)
        for _ in range(50):
            state, reward, terminal, _ = env.step(command)

        # now introduce random perturbations
        p = [] #past pixel snapshots
        WINDOW_SIZE = 2
        for i in range(WINDOW_SIZE):
            p.append(np.zeros([512,512,3]))
        p = np.array(p)

        velocity = state[Sensors.VELOCITY_SENSOR]
        for i in range(500):
            #command = np.array([0, 0, 0, 10])
            raw_command = reinforce.select_action(np.squeeze(velocity)/1000)
            command = get_vector(raw_command)
            #print(np.hstack([np.squeeze(velocity),command]))
            state, reward, terminal, _ = env.step(command)

            # To access specific sensor data:
            pixels = state[Sensors.PRIMARY_PLAYER_CAMERA]
            velocity = state[Sensors.VELOCITY_SENSOR]
            orientation = state[Sensors.ORIENTATION_SENSOR]
            location = state[Sensors.LOCATION_SENSOR] #Not sure this is working
            imu = state[Sensors.IMU_SENSOR] #Not sure this is working
            # For a full list of sensors the UAV has, view the README

            velocity_cost = np.sum(abs(velocity))
            avg_velocity_cost += velocity_cost

            reinforce.policy.rewards.append(-1 * velocity_cost)

            #if i%50 == 0:
            #    print('\n')
            #    print(i)
            #    print(command)
            #    print(velocity_cost)

            #print("VECLOCITY")
            #print(velocity)
            #print("IMU")
            #print(imu)

            for i in reversed(range(WINDOW_SIZE)):
                if i > 0:
                    p[i] = p[i-1]
            p[0] = pixels[:,:,[2,1,0]]

            if True:
                deltas = 255*(abs(np.mean(p[-1:,:,:,:],axis=0)-np.mean(p[:-1,:,:,:], axis=0))>50)
                pixel_cost = np.sum(deltas)
                delta_surface = pygame.surfarray.make_surface(deltas)

                #print(velocity_cost)
                #print(str(pixel_cost) + '\t' + str(velocity_cost))

                d2 = pygame.transform.rotate(delta_surface,-90)
                d3 = pygame.transform.flip(d2,1,0)
    
                if HEADLESS != True:
                    screen.blit(d3, (0,0))
                    pygame.display.flip()
           
        #parse the reinforcement learning rewards
        reinforce.finish_episode()
        print("EPISODE %d\t%f"%(e,avg_velocity_cost/NUM_TRIALS))

        if e%100 == 0:
            print("Saving model...")
            pickle.dump(reinforce.policy,open('drone_policies/' + str(e) + '.pkl','wb'))

if __name__ == "__main__":
    pygame.init()
    size = width, height = 512, 512
    speed = [2, 2]
    black = 0, 0, 0

    if HEADLESS != True:
        screen = pygame.display.set_mode(size)
    else:
        screen = None

    learn_to_hold_position(screen)
    print("Finished")
